JoinColony/colonyDapp
# Create Initiate Safe Transaction Saga

## Description

This PR introduces the saga that fires when the Gnosis Control Safe form is submitted. It is extended in #3875 to include NFT Transfer.

You must have the foreign chain and bridge running in order for this to work (see instructions below).
 
## To Test Locally

1. Have the foreign chain and bridge env running (see intstructions below)
1. Set the following .env variables to the addresses of the mock deployments. This information will be logged to the console in the window in which you're running the foreign chain (it changes every time you restart the bridge).
`ZODIAC_BRIDGE_MODULE_ADDRESS= Zodiac Bridge module address`
`LOCAL_FOREIGN_BRIDGE_ADDRESS= Foreign bridge address`
`LOCAL_HOME_BRIDGE_ADDRESS= Home bridge address`
1. After setting your .env variables, restart webpack.
1. Navigate to the Control Safe dialog.
1. Select a safe. Currently doesn't matter which (in local env) as it'll be using the zodiac module bridge which is already connected to the safe deployed locally.
1. Select raw transaction as the transaction type.
1. The recipient can be anybody.
1. Enter  an amount. 
1. Enter byte data beginning with "0x". If in doubt, just enter "0x".
1. Repeat for as many transactions as you wish.
1. Submit the form. 
1. The gas station popover should pop up with the annotation action. You should then be redirected to the ActionsPage (which is yet to be wired -- #3821). You can comment out the redirect in the saga to see this.
1. You should see printed in the console a message that contains `true` -- this tells you that the `executeTransaction` function was successfully executed on the zodiac module bridge. 

**New stuff** ✨

* Safe Transaction Saga
* safeHelpers.ts 
 
**Changes** 🏗

* connected `GnosisSafeControlDialog` to saga
* Set up `AnnotateTransaction` as a processable event in the `colonyAction` resolver, so that we can be redirected to the action page. 

## Setup Bridge instructions

0. have your normal dev env running. 
1. `cd src/lib/colonyNetwork` 
3. `git fetch` 
4. `git checkout maint/cross-chain-safe` 
5. `git submodule update --init` 
6. `yarn` 
7. `yarn provision:safe:contracts` 
8. `cd lib/safe-contracts` then `yarn`  (ignore the errors)
9. `cd ../..` (back into colonyNetwork) 
10. `bash scripts/start-bridging-environment.sh` 
11. `node scripts/start-bridging-environment.js` 

## Still Todo in other issues

- Add logic for all other transaction types
- confirm how we're going to handle this in production...

Resolves #3721
